pkg/controller/scheduler.go

```diff
@@ -398,7 +401,7 @@ func (c *Controller) advanceCanary(name string, namespace string) {
 
 	// check if the canary success rate is above the threshold
 	// skip check if no traffic is routed or mirrored to canary
-	if canaryWeight == 0 && cd.Status.Iterations == 0 &&
+	if cd.Status.Phase == flaggerv1.CanaryPhaseProgressing && canaryWeight == 0 && cd.Status.Iterations == 0 &&
```

Comment:
Is this intentional? Why would we want to assert that `cd.Status.Phase` is `CanaryPhaseProgressing` when we want to start the canary analysis?

Reply:
@aryan9600 Yes this is intentional. This is actually what I'm intending to fix by adding a dedicated phase `WaitingIncreaseTraffic`. Previously, when it calls `runConfirmTrafficIncreaseHooks` and the webhook is failed, it returns and reaches this code you highlight it satisfies the condition canaryWeight == 0, Status.Iterations == 0. So, it's restarted to preRollout again.

In this PR, it's set to WaitingTrafficIncrease and we limit this code to CanaryPhaseProgressing and this fixes the bug by preventing WaitingTrafficIncrease to enter this code block.

Reply:
Yes, but this'd break the existing state machine. The canary `Phase` is set to `Progressing` after the canary analysis starts, so if we try to assert that the `Phase` is `Progressing` here, the canary analysis will never start.

Reply:
The canary Phase is set to Progressing inside this function `checkCanaryStatus`, which happens before the code we are discussing.

```go
	// check if canary analysis should start (canary revision has changes) or continue
	if ok := c.checkCanaryStatus(cd, canaryController, scalerReconciler, shouldAdvance); !ok {
		return
	}
```

Also before making this change, I have debugged the Phase when it enters this code block, and it has Phase: Progressing.

In my test, this is working, the canary analysis starts and it works as expected. The e2e tests also passed for this PR before I added more commit through GitHub that turns out missing the Signed-off message.